readme.md
# Adaptive Images

Is a solution to automatically create, cache, and deliver device-appropriate versions of your website's content images. 
It does not require you to change your mark-up. 
It is intended for use with [Responsive Designs](http://www.abookapart.com/products/responsive-web-design) and to be combined with [Fluid Image](http://unstoppablerobotninja.com/entry/fluid-images/) techniques.

## How To

Copy `adaptive-images.php`, `adaptive-images-config.php` and `.htaccess` into the root directory of your site. 
If you already have a `.htaccess` file __DO NOT OVERWRITE IT__, skip down to the advanced instructions.

You can modify the configuration options in the `adaptive-images-config.php` file.

The `adaptive-images.php` script will create an `ai-cache` folder in the location specified inside the configuration file.  
If you are extra paranoid about security you can have the ai-cache directory sit outside of your web-root so it's not publicly accessible. 
Just set the paths properly in the `.htaccess` file and the `.php` configuration file.

You do not need the `adaptive-images-cookie.php` file unless you are using the __alternate method__ (see below) of detecting the users screen size. 
_So delete it if you like, no one likes mess._


### Update your `.htacces`
*You already have a .htaccess file*
I strongly advise you to duplicate that file so you can revert to it if things go pear-shaped.
Open your existing `.htaccess` file and edit the contents. 
You'll need to look and see if there is a section that begins with the following:

    <IfModule mod_rewrite.c>

If there is, then you need to add the following lines into that block:

    # Adaptive-Images -----------------------------------------------------------------------------------
    # Add any directories you wish to omit from the Adaptive-Images process on a new line, as follows:
    # RewriteCond %{REQUEST_URI} !some-directory
    # RewriteCond %{REQUEST_URI} !another-directory
    RewriteCond %{REQUEST_URI} !assets
    # Send any GIF, JPG, or PNG request that IS NOT stored inside one of the above directories
    # to adaptive-images.php so we can select appropriately sized versions
    RewriteRule \.(?:jpe?g|gif|png)$ adaptive-images.php
    # END Adaptive-Images -------------------------------------------------------------------------------

Otherwise, the complete `<IfModule mod_rewrite.c>` block in your `.htaccess` should look like this

    <IfModule mod_rewrite.c>
      Options +FollowSymlinks
      RewriteEngine On
      # Adaptive-Images -----------------------------------------------------------------------------------
      # Add any directories you wish to omit from the Adaptive-Images process on a new line, as follows:
      # RewriteCond %{REQUEST_URI} !some-directory
      # RewriteCond %{REQUEST_URI} !another-directory
      RewriteCond %{REQUEST_URI} !assets
      # Send any GIF, JPG, or PNG request that IS NOT stored inside one of the above directories
      # to adaptive-images.php so we can select appropriately sized versions
      RewriteRule \.(?:jpe?g|gif|png)$ adaptive-images.php
      # END Adaptive-Images -------------------------------------------------------------------------------
    </IfModule>

_Instructions are in the file as comments (any line that starts with a # is a comment, and doesn't actually do anything)_

### Update your `<head>`

Copy the following Javascript into the `<head>` of your site. 
It MUST go in the head as the first bit of JS, before any other JS. 
This is because it needs to work as soon as possible, any delay wil have adverse effects.

    <script>document.cookie='resolution='+Math.max(screen.width,screen.height)+'; path=/';</script>


If you would like to take advantage of high-pixel density devices such as the iPhone4 or iPad3 Retina display you can use the following JavaScript instead. 
It will send higher-resolution images to such devices - be aware this will mean slower downloads for Retina users, but better images.

    <script>document.cookie='resolution='+Math.max(screen.width,screen.height)+("devicePixelRatio" in window ? ","+devicePixelRatio : ",1")+'; path=/';</script>

### Configurations

You can now open the `adaptive-images-config.php` file and have a play with the settings that are in the CONFIG area. By default it looks like this:

    /* CONFIG ----------------------------------------------------------------------------------------------------------- */
    define('AI_COOKIE_NAME', 'resolution');
    $resolutions   = array(1382, 992, 768, 480); // the resolution break-points to use (screen widths, in pixels)
    $cache_path    = dirname($_SERVER["PHP_SELF"])."/ai-cache"; // where to store the generated re-sized images. From your document root!
    $jpg_quality   = 75; // the quality of any generated JPGs on a scale of 0 to 100
    $sharpen       = TRUE; // Shrinking images can blur details, perform a sharpen on re-scaled images?
    $watch_cache   = TRUE; // check that the adapted image isn't stale (ensures updated source images are re-cached)
    $browser_cache = 60*60*24*7; // How long the BROWSER cache should last (seconds, minutes, hours, days. 7days by default)
    /* END CONFIG ------------------------------------------------------------------------------------------------------- */


-   `$resolutions` are the screen widths we'll work with. 
In the default it will store a re-sized image for large screens, normal screens, tablets, phones, and tiny phones.
In general, if you're using a responsive design in CSS, these breakpoints will be exactly the same as the ones you use in your media queries.

-   `$cache_path` If you don't like the cached images being written to that folder, you can put it somewhere else.
Just put the path to the folder here and make sure you create it on the server if for some reason that couldn't be done autmoatically by `adaptive-images.php`.

-    `$sharpen` Will perform a subtle sharpening on the rescaled images.
Usually this is fine, but you may want to turn it off if your server is very very busy.

-    `$watch_cache` If your server gets very busy it may help performance to turn this to FALSE. 
It will mean however that you will have to manually clear out the cache directory if you change a resource file


## Alternate method for those who can't rely on JavaScript

One of the weaknesses of the Adaptive Images process is its reliance on JavaScript to detect the visitors screen size, and set a cookie with that screen size inside it. 
The following is a solution for people who need the system to work without the use of JavaScript, but it does have a major caveat which is why this is not the default method, and why it is "unsupported" (I'm not going to troubleshoot problems you have with this method).

### The alternative method

__Into the head__, _above_ your normal CSS `<link>` tags, add the following:

    <noscript><style>
       @media only screen and (max-device-width: 479px) {
         html { background-image:url(adaptive-images-cookie.php?maxwidth=479); }
       }
       @media only screen and (min-device-width: 480px) and (max-device-width: 767px) {
         html { background-image:url(adaptive-images-cookie.php?maxwidth=767); }
       }
       @media only screen and (min-device-width: 768px) and (max-device-width: 991px) {
         html { background-image:url(adaptive-images-cookie.php?maxwidth=991); }
       }
       @media only screen and (min-device-width: 992px) and (max-device-width: 1381px) {
         html { background-image:url(adaptive-images-cookie.php?maxwidth=1381); }
       }
       @media only screen and (min-device-width: 1382px) {
         html { background-image:url(adaptive-images-cookie.php?maxwidth=unknown); }
       }
    </style></noscript>

If you use this method you will need to ensure the widths here match those in your `adaptive-images.php` file, you will also need to have the `adaptive-images-cookie.php` file in the root of your website (_no one else needs this file_).

### The caveat

Using this method additionally (or instead) of the JS method makes it likely that on the very first visit to your site, the images sent to the visitor will be the original full-scale versions. 
However, __ALL__ subsequent pages on your site will work properly. 
What that means is that, really, this solution is only viable if you've got `$mobile_first` set to FALSE. 
Otherwise, the majority of people who visit your site will experience images too small for their computer on the very first visit.

The reason is to do with how browsers load web pages. 
The first thing a browser does is load the HTML, in the order it's written - so for a normal AI install it loads the HTML and see's the embeded JavaScript and immediately executes that JavaScript - which writes the cookie. 
It then carries on loading the rest of the page in the order it finds it. 
Long story short - it means that when it finds an image tag and asks the server for it, it already has a cookie stored.

That's not likely to be the case if you use the CSS method. 
Because the CSS method relies on an external file - it has to ask the server for the "`background-image`", which is really just a bit of PHP to set a cookie. 
The problem is that when a browser loads external files like that, it doesn't stop loading the HTML, it carries on doing that at the same time as waiting for the external file.
Which means that it can get to an image tag before the server has finished loading that PHP file. 
This is only an issue on the very first page load, and AI has smart fallbacks to deal with a no-cookie situation.

## Troubleshooting

Most of the time people report problems it is due to one of two things:

1.   If images vanish, there is something wrong with your `.htaccess` configuration. 
This happens mostly on WordPress sites - it's because the server, and wordpress, have specific requirements that are different from most servers. 
You'll have to play about in the `.htaccess` file and read up on how to use `ModRewrite`.

2.   If you're seeing error images (big black ones) that's AI working, so your `.htaccess` is fine.
Read the messages on the image. 
Most of the time you'll only see this problem because your server requires less strict permissions to write to the disc. 
Try setting the `ai-cache` directory to `775`, and if all else fails use `777` - but be aware this is not very secure if you're on a shared server, and you ought to instead contact your administrator to get the server set up properly.


## Benefits

- Ensures you and your visitors are not wasting bandwidth delivering images at a higher resolution than the vistor needs.
- Will work on your existing site, as it requires no changes to your mark-up.
- Is device agnostic (it works by detecting the size of the visitors screen)
- Is CMS agnostic (it manages its own image re-sizing, and will work on any CMS or even on flat HTML pages)
- Is entirely automatic. Once added to your site, you need do no further work.
- Highly configurable
    - Set the resolutions you want to become adaptive (usually the same as the ones in your CSS @media queries)
    - Choose where you want the cached files to be stored
    - Configure directories to ignore (protect certain directories so AI is not applied to images within them)

Find out more, and view examples at [Adaptive Images](http://adaptive-images.com)

## Legal

Adaptive Images by Matt Wilcox is licensed under a [Creative Commons Attribution 3.0 Unported License](http://creativecommons.org/licenses/by/3.0)


changelog.txt
version 1.5.3 (2013/05/13) ------------------------------------------------------------

  EXPERIMENTAL These edits have not been tested YET
  ENHANCEMENT
    * Personalized name of the COOKIE
    * Unified configuration in a separate conf file
    * ai-cache directory path is now using $_SERVER["PHP_SELF"]
    * Unified file names
    * Moved all instructions in the readme
  
version 1.5.2 (2012/04/16) ------------------------------------------------------------

  BUG FIX
    * AI was not generating high DPI images larger than the existing maximum
      $resolutions value
      
      I used a < when I needed a >. That is now fixed. If AI detects retina displays it
      is now able to generate larger images than the top bounds set in $resolutions.

version 1.5.1 (2012/04/14) ------------------------------------------------------------

  ENHANCEMENT
    * JPGs are now created as Progressive rather than Optimised.
      
      Progressive JPGs are actually slightly smaller in file-size than Optimised JPGs,
      and have the benefit of appearing to load faster in modern browsers (they
      download in passes, gradually getting sharper).

version 1.5 (2012/04/14) --------------------------------------------------------------

  ENHANCEMENT
    * Much improved support for Retina and other high DPI devices

      NOTE: Retina support requires using alternate JavaScript:
      <script>document.cookie='resolution='+Math.max(screen.width,screen.height)+("devicePixelRatio" in window ? ","+devicePixelRatio : ",1")+'; path=/';</script>

      NOTE: If you see folders in the ai-cache directory which are *larger* than any
      defined in your $resolutions array, it is because someone with a high DPI / Retina
      display visited your site, and AI was able to generate the required image for them.

      DETAILS:
      Previous AI behaviour meant that supporting Retina required adding very high values
      in the $resolutions array. This is sub-optimal because:

      1) This array is meant to be nothing more than the same values as your media query
         breakpoints. Adding larger ones isn't intuitive.
      2) When no cookie is set, AI sends non-mobile devices the largest $resolution value,
         which in that case could be far too large for most people (2000px+)

      AI is now much smarter, you do not need to edit the $resolutions array; just leave
      that as your CSS MQ sizes.

      AI now auto-detects the pixel density of any high DPI device and either picks a
      matching size from the existing array, or creates new high-dpi images which are
      multiples of your chosen $resolutions values.

version 1.4.1 (2012/03/28) ------------------------------------------------------------

  ENHANCEMENT
    * More helpful error messages if something goes wrong.
    * Updated the documentation a little.

version 1.4 (2012/02/19) --------------------------------------------------------------

  ENHANCEMENT
    * Auto-creates the cache folder with the correct permissions.
      No need to manually create the ai-cache folder any more, AI does it for you.

  BUG FIX
    * Fixed an incorrect permission on created files, they're now 755 not 777.

version 1.3.9 (2012/02/17) ------------------------------------------------------------
                                                                         Thanks to Gerv

  ENHANCEMENT
    * Better detection of mobile device in the event cookies are not available.
      
      If cookies are unavailable AI resorts to basic UA sniffing.

      Previously AI assumed all devices are mobile and searched the UA string for a
      desktop OS. If found it overrode $mobile_first. However, this meant tablet
      devices got mobile sized images as their OS is not a desktop OS.

      Now switched to assume the device is a desktop, but AI searches the UA string for
      'mobile' to switch to mobile first. All current mobile browsers behave in this
      way, meaning mobiles still get mobile resolution, but tablets and desktop get
      the largest configured resolution.

    * Removed $mobile_first as it is no longer needed (see above)

version 1.3.8 (2012/01/24) ------------------------------------------------------------

  BUG FIX
    * Eliminated occasional double-slash in $cache_file path due to $requested_uri
      sometimes starting with a /

version 1.3.7 (2011/11/21) ------------------------------------------------------------

  BUG FIX
    * A stupid mistake was causing the browser detect to always return false.
      This has been fixed.

      With this update Adaptive Images now detects if a user is running on a desktop OS
      and switches $mobile_first to FALSE in that case. This means that if no cookie is
      set and $mobile_first is TRUE the following will happen:

      On a mobile phone, the mobile sized image will be delivered.
      On a desktop, the highest resolution configured will be delivered.

      If a cookie is set, Adaptive Images will use the cookie value to determine the
      correct size image to send.

version 1.3.6 (2011/11/11) ------------------------------------------------------------
  
  NOTES
    * Added Linux desktop detection to the $mobile_first check routine (see previous
      changelog details)

version 1.3.5 (2011/11/10) ------------------------------------------------------------
  
  NEW FEATURE
    * To work around browser issues with unreliable cookie setting speed
      Adaptive Images now also checks the user agent string.
      
      If a desktop environment is detected, $mobile_first is automatically
      over-ridden to "FALSE". If not Adaptive Images will obey the configured
      value.
      
      This is a safety fallback and requires testing before I revert the
      default $mobile_first configured value to TRUE.

version 1.3.4 (2011/11/10) ------------------------------------------------------------
  
  BUG FIXES

adaptive-images.php
<?php
/* PROJECT INFO --------------------------------------------------------------------------------------------------------
   Version:   1.5.3
   Changelog: http://adaptive-images.com/changelog.txt

   Homepage:  http://adaptive-images.com
   GitHub:    https://github.com/MattWilcox/Adaptive-Images
   Twitter:   @responsiveimg

   Edited by: Kuzeko

   LEGAL:
   Adaptive Images by Matt Wilcox is licensed under a Creative Commons Attribution 3.0 Unported License.
--------------------------------------------------------------------------------------------------------------------- */

define('AI_VERSION', '1.5.3');

/* INCLUDE CONFIG --------------------------------------------------------------------------------------------------- */

require_once('adaptive-images-config.php');

/* ---------------------------------------------------------------------------------------------------------------------
------------------------ Don't edit anything after this line unless you know what you're doing -------------------------
--------------------------------------------------------------------------------------------------------------------- */

/* get all of the required data from the HTTP request */
$document_root  = $_SERVER['DOCUMENT_ROOT'];
$requested_uri  = parse_url(urldecode($_SERVER['REQUEST_URI']), PHP_URL_PATH);
$requested_file = basename($requested_uri);
$source_file    = $document_root.$requested_uri;
$resolution     = FALSE;

/* Mobile detection
   NOTE: only used in the event a cookie isn't available. */
function is_mobile() {
  $userAgent = strtolower($_SERVER['HTTP_USER_AGENT']);
  return !(strpos($userAgent, 'mobile') === FALSE);
}

/* Does the UA string indicate this is a mobile? */
$is_mobile = is_mobile();


// does the $cache_path directory exist already?
if (!is_dir("$document_root/$cache_path")) { // no
  if (!mkdir("$document_root/$cache_path", 0755, true)) { // so make it
    if (!is_dir("$document_root/$cache_path")) { // check again to protect against race conditions
      // uh-oh, failed to make that directory
      sendErrorImage("Failed to create cache directory at: $document_root/$cache_path");
    }
  }
}

/* helper function: Send headers and returns an image. */
function sendImage($filename, $browser_cache) {
  $extension = strtolower(pathinfo($filename, PATHINFO_EXTENSION));
  if (in_array($extension, array('png', 'gif', 'jpeg'))) {
    header("Content-Type: image/".$extension);
  } else {
    header("Content-Type: image/jpeg");
  }
  header("Cache-Control: private, max-age=".$browser_cache);
  header('Expires: '.gmdate('D, d M Y H:i:s', time()+$browser_cache).' GMT');
  header('Content-Length: '.filesize($filename));
  readfile($filename);
  exit();
}

/* helper function: Create and send an image with an error message. */
function sendErrorImage($message) {
  /* get all of the required data from the HTTP request */
  $document_root  = $_SERVER['DOCUMENT_ROOT'];
  $requested_uri  = parse_url(urldecode($_SERVER['REQUEST_URI']), PHP_URL_PATH);
  $requested_file = basename($requested_uri);
  $source_file    = $document_root.$requested_uri;

  if(!is_mobile()){
    $is_mobile = "FALSE";
  } else {
    $is_mobile = "TRUE";
  }

  $im            = ImageCreateTrueColor(800, 300);
  $text_color    = ImageColorAllocate($im, 233, 14, 91);
  $message_color = ImageColorAllocate($im, 91, 112, 233);

  ImageString($im, 5, 5, 5, "Adaptive Images encountered a problem:", $text_color);
  ImageString($im, 3, 5, 25, $message, $message_color);

  ImageString($im, 5, 5, 85, "Potentially useful information:", $text_color);
  ImageString($im, 3, 5, 105, "DOCUMENT ROOT IS: $document_root", $text_color);
  ImageString($im, 3, 5, 125, "REQUESTED URI WAS: $requested_uri", $text_color);
  ImageString($im, 3, 5, 145, "REQUESTED FILE WAS: $requested_file", $text_color);
  ImageString($im, 3, 5, 165, "SOURCE FILE IS: $source_file", $text_color);
  ImageString($im, 3, 5, 185, "DEVICE IS MOBILE? $is_mobile", $text_color);

  header("Cache-Control: no-store");
  header('Expires: '.gmdate('D, d M Y H:i:s', time()-1000).' GMT');
  header('Content-Type: image/jpeg');
  ImageJpeg($im);
  ImageDestroy($im);
  exit();
}

/* sharpen images function */
function findSharp($intOrig, $intFinal) {
  $intFinal = $intFinal * (750.0 / $intOrig);
  $intA     = 52;
  $intB     = -0.27810650887573124;
  $intC     = .00047337278106508946;
  $intRes   = $intA + $intB * $intFinal + $intC * $intFinal * $intFinal;
  return max(round($intRes), 0);
}

/* refreshes the cached image if it's outdated */
function refreshCache($source_file, $cache_file, $resolution) {
  if (file_exists($cache_file)) {
    // not modified
    if (filemtime($cache_file) >= filemtime($source_file)) {
      return $cache_file;
    }

    // modified, clear it
    unlink($cache_file);
  }
  return generateImage($source_file, $cache_file, $resolution);
}

/* generates the given cache file for the given source file with the given resolution */
function generateImage($source_file, $cache_file, $resolution) {
  global $sharpen, $jpg_quality;

  $extension = strtolower(pathinfo($source_file, PATHINFO_EXTENSION));

  // Check the image dimensions
  $dimensions   = GetImageSize($source_file);
  $width        = $dimensions[0];
  $height       = $dimensions[1];

  // Do we need to downscale the image?
  if ($width <= $resolution) { // no, because the width of the source image is already less than the client width
    return $source_file;
  }

  // We need to resize the source image to the width of the resolution breakpoint we're working with
  $ratio      = $height/$width;
  $new_width  = $resolution;
  $new_height = ceil($new_width * $ratio);
  $dst        = ImageCreateTrueColor($new_width, $new_height); // re-sized image

  switch ($extension) {
    case 'png':
      $src = @ImageCreateFromPng($source_file); // original image
    break;
    case 'gif':
      $src = @ImageCreateFromGif($source_file); // original image
    break;
    default:
      $src = @ImageCreateFromJpeg($source_file); // original image
      ImageInterlace($dst, true); // Enable interlancing (progressive JPG, smaller size file)
    break;
  }

  if($extension=='png'){
    imagealphablending($dst, false);
    imagesavealpha($dst,true);
    $transparent = imagecolorallocatealpha($dst, 255, 255, 255, 127);
    imagefilledrectangle($dst, 0, 0, $new_width, $new_height, $transparent);
  }

  ImageCopyResampled($dst, $src, 0, 0, 0, 0, $new_width, $new_height, $width, $height); // do the resize in memory
  ImageDestroy($src);

  // sharpen the image?
  // NOTE: requires PHP compiled with the bundled version of GD
  // (see http://php.net/manual/en/function.imageconvolution.php)
  if($sharpen == TRUE && function_exists('imageconvolution')) {
    $intSharpness = findSharp($width, $new_width);
    $arrMatrix = array(
      array(-1, -2, -1),
      array(-2, $intSharpness + 12, -2),
      array(-1, -2, -1)
    );
    imageconvolution($dst, $arrMatrix, $intSharpness, 0);
  }

  $cache_dir = dirname($cache_file);

  // does the directory exist already?
  if (!is_dir($cache_dir)) {
    if (!mkdir($cache_dir, 0755, true)) {
      // check again if it really doesn't exist to protect against race conditions
      if (!is_dir($cache_dir)) {
        // uh-oh, failed to make that directory
        ImageDestroy($dst);
        sendErrorImage("Failed to create cache directory: $cache_dir");
      }
    }
  }

  if (!is_writable($cache_dir)) {
    sendErrorImage("The cache directory is not writable: $cache_dir");
  }

  // save the new file in the appropriate path, and send a version to the browser
  switch ($extension) {
    case 'png':
      $gotSaved = ImagePng($dst, $cache_file);
    break;
    case 'gif':
      $gotSaved = ImageGif($dst, $cache_file);
    break;
    default:
      $gotSaved = ImageJpeg($dst, $cache_file, $jpg_quality);
    break;
  }
  ImageDestroy($dst);

  if (!$gotSaved && !file_exists($cache_file)) {
    sendErrorImage("Failed to create image: $cache_file");
  }

  return $cache_file;
}

// check if the file exists at all
if (!file_exists($source_file)) {
  header("Status: 404 Not Found");
  exit();
}

/* check that PHP has the GD library available to use for image re-sizing */
if (!extension_loaded('gd')) { // it's not loaded
  if (!function_exists('dl') || !dl('gd.so')) { // and we can't load it either
    // no GD available, so deliver the image straight up
    trigger_error('You must enable the GD extension to make use of Adaptive Images', E_USER_WARNING);
    sendImage($source_file, $browser_cache);
  }
}

/* Check to see if a valid cookie exists */
if (isset($_COOKIE[AI_COOKIE_NAME])) {
  $cookie_value = $_COOKIE[AI_COOKIE_NAME];

  // does the cookie look valid? [whole number, comma, potential floating number]
  if (! preg_match("/^[0-9]+[,]*[0-9\.]+$/", "$cookie_value")) { // no it doesn't look valid
    setcookie("resolution", "$cookie_value", time()-100); // delete the mangled cookie
  }
  else { // the cookie is valid, do stuff with it
    $cookie_data   = explode(",", $_COOKIE[AI_COOKIE_NAME]);
    $client_width  = (int) $cookie_data[0]; // the base resolution (CSS pixels)
    $total_width   = $client_width;
    $pixel_density = 1; // set a default, used for non-retina style JS snippet
    if (@$cookie_data[1]) { // the device's pixel density factor (physical pixels per CSS pixel)
      $pixel_density = $cookie_data[1];
    }

    rsort($resolutions); // make sure the supplied break-points are in reverse size order
    $resolution = $resolutions[0]; // by default use the largest supported break-point

    // if pixel density is not 1, then we need to be smart about adapting and fitting into the defined breakpoints
    if($pixel_density != 1) {
      $total_width = $client_width * $pixel_density; // required physical pixel width of the image

      // the required image width is bigger than any existing value in $resolutions
      if($total_width > $resolutions[0]){
        // firstly, fit the CSS size into a break point ignoring the multiplier
        foreach ($resolutions as $break_point) { // filter down
          if ($total_width <= $break_point) {
            $resolution = $break_point;
          }
        }
        // now apply the multiplier
        $resolution = $resolution * $pixel_density;
      }
      // the required image fits into the existing breakpoints in $resolutions
      else {
        foreach ($resolutions as $break_point) { // filter down
          if ($total_width <= $break_point) {
            $resolution = $break_point;
          }
        }
      }
    }
    else { // pixel density is 1, just fit it into one of the breakpoints
      foreach ($resolutions as $break_point) { // filter down
        if ($total_width <= $break_point) {
          $resolution = $break_point;
        }
      }
    }
  }
}

/* No resolution was found (no cookie or invalid cookie) */
if (!$resolution) {
  // We send the lowest resolution for mobile-first approach, and highest otherwise
  $resolution = $is_mobile ? min($resolutions) : max($resolutions);
}

/* if the requested URL starts with a slash, remove the slash */
if(substr($requested_uri, 0,1) == "/") {
  $requested_uri = substr($requested_uri, 1);
}

/* whew might the cache file be? */
$cache_file = $document_root."/$cache_path/$resolution/".$requested_uri;

/* Use the resolution value as a path variable and check to see if an image of the same name exists at that path */
if (file_exists($cache_file)) { // it exists cached at that size
  /* if cache watching is enabled, compare cache and source modified dates to ensure the cache isn't stale */
  if ($watch_cache) {
    $cache_file = refreshCache($source_file, $cache_file, $resolution);
  }

  sendImage($cache_file, $browser_cache);
}

/* It exists as a source file, and it doesn't exist cached - lets make one: */
$file = generateImage($source_file, $cache_file, $resolution);
sendImage($file, $browser_cache);


adaptive-images-cookie.php
<?php
/* PROJECT INFO --------------------------------------------------------------------------------------------------------
   Version:   1.5.3
   Changelog: http://adaptive-images.com/changelog.txt

   Homepage:  http://adaptive-images.com
   GitHub:    https://github.com/MattWilcox/Adaptive-Images
   Twitter:   @responsiveimg

   Edited by: Kuzeko

   LEGAL:
   Adaptive Images by Matt Wilcox is licensed under a Creative Commons Attribution 3.0 Unported License.
--------------------------------------------------------------------------------------------------------------------- */

/* EXAMPLE -------------------------------------------------------------------------------------------------------------
/* Put this EXAMPLE up in the head of your <html> if you do not want to rely on Javascript -----------------------------

<noscript><style>
   @media only screen and (max-device-width: 479px) {
     html { background-image:url(adaptive-images-cookie.php?maxwidth=479); }
   }
   @media only screen and (min-device-width: 480px) and (max-device-width: 767px) {
     html { background-image:url(adaptive-images-cookie.php?maxwidth=767); }
   }
   @media only screen and (min-device-width: 768px) and (max-device-width: 991px) {
     html { background-image:url(adaptive-images-cookie.php?maxwidth=991); }
   }
   @media only screen and (min-device-width: 992px) and (max-device-width: 1381px) {
     html { background-image:url(adaptive-images-cookie.php?maxwidth=1381); }
   }
   @media only screen and (min-device-width: 1382px) {
     html { background-image:url(adaptive-images-cookie.php?maxwidth=unknown); }
   }
</style></noscript>
/* END EXAMPLE ------------------------------------------------------------------------------------------------------ */

/* CONFIG ----------------------------------------------------------------------------------------------------------- */

define('AI_VERSION', '1.5.3');
require_once('adaptive-images-config.php');

/* END CONFIG ------------------------------------------------------------------------------------------------------- */

// we need a number, so if it is undefined, give it something much bigger than the maximum as configured
$maxwidth = isset($_GET['maxwidth']) && is_numeric($_GET['maxwidth'])? intval($_GET['maxwidth']) : $resolutions[0]*2;

setcookie(COOKIE_NAME,$maxwidth,time()+$browser_cache,'/'); // set the cookie

// respond with an empty content
header('HTTP/1.1 204 No Content');
exit();


adaptive-images-config.php
<?php
/* If this is requested directly, no content should be submitted */
if(!defined(AI_VERSION)){
    // respond with not found
    header('HTTP/1.1 404 Not Found');
    exit();
}

/* CONFIG ----------------------------------------------------------------------------------------------------------- */

define('AI_COOKIE_NAME', 'resolution');

$resolutions   = array(1382, 992, 768, 480); // the resolution break-points to use (screen widths, in pixels)
$cache_path    = dirname($_SERVER["PHP_SELF"])."/ai-cache"; // where to store the generated re-sized images. From your document root!
$jpg_quality   = 75; // the quality of any generated JPGs on a scale of 0 to 100
$sharpen       = TRUE; // Shrinking images can blur details, perform a sharpen on re-scaled images?
$watch_cache   = TRUE; // check that the adapted image isn't stale (ensures updated source images are re-cached)
$browser_cache = 60*60*24*7; // How long the BROWSER cache should last (seconds, minutes, hours, days. 7days by default)

/* END CONFIG ------------------------------------------------------------------------------------------------------- */
